File: src/Player.js
export class Player {
  /**
   * @param {string} id
   * @param {string} name
   * @param {{active: import("./Card.js").Card[], reserve: import("./Card.js").Card[]}} deck
   */
  constructor(id, name, deck) {
    this.id = id;
    this.name = name;

    // Active slots are a fixed-size array; a defeated card leaves an
    // `null` hole until the player swaps a reserve into it.
    this.activeSlots = [...deck.active];
    this.reserve = [...deck.reserve];
    this.graveyard = [];
  }

  get allCards() {
    return [...this.activeSlots.filter(Boolean), ...this.reserve, ...this.graveyard];
  }

  get livingActiveCards() {
    return this.activeSlots.filter((c) => c && c.isAlive);
  }

  get livingReserveCards() {
    return this.reserve.filter((c) => c.isAlive);
  }

  /** True once every card the player owns (active + reserve) is defeated. */
  get hasLost() {
    const activeAlive = this.activeSlots.some((c) => c && c.isAlive);
    const reserveAlive = this.reserve.some((c) => c.isAlive);
    return !activeAlive && !reserveAlive;
  }

  /** Move a card from reserve into an empty active slot. */
  swapIntoActive(reserveIndex, activeSlotIndex) {
    const reserveCard = this.reserve[reserveIndex];
    if (!reserveCard) throw new Error("No card at that reserve index.");
    if (!reserveCard.isAlive) throw new Error("Can't field a defeated card.");
    if (this.activeSlots[activeSlotIndex]) {
      throw new Error("That active slot isn't empty.");
    }

    this.reserve.splice(reserveIndex, 1);
    reserveCard.zone = "active";
    this.activeSlots[activeSlotIndex] = reserveCard;
    return reserveCard;
  }

  /** Called after a card's HP hits 0: clears its slot, moves it to the graveyard. */
  handleDefeat(card) {
    const activeIdx = this.activeSlots.indexOf(card);
    if (activeIdx !== -1) {
      this.activeSlots[activeIdx] = null;
    }
    const reserveIdx = this.reserve.indexOf(card);
    if (reserveIdx !== -1) {
      this.reserve.splice(reserveIdx, 1);
    }
    card.zone = "graveyard";
    this.graveyard.push(card);
  }

  hasEmptyActiveSlot() {
    return this.activeSlots.some((c) => c === null);
  }
}

File: game/player.py
class Player:

    def __init__(self, id, name, deck):
        self.id = id
        self.name = name

        self.active_slots = list(deck["active"])
        self.reserve = list(deck["reserve"])
        self.graveyard = []

    @property
    def all_cards(self):
        return (
            [
                card
                for card in self.active_slots
                if card is not None
            ]
            + self.reserve
            + self.graveyard
        )

    @property
    def living_active_cards(self):
        return [
            card
            for card in self.active_slots
            if card is not None and card.is_alive
        ]

    @property
    def living_reserve_cards(self):
        return [
            card
            for card in self.reserve
            if card.is_alive
        ]

    @property
    def has_lost(self):
        active_alive = any(
            card is not None and card.is_alive
            for card in self.active_slots
        )

        reserve_alive = any(
            card.is_alive
            for card in self.reserve
        )

        return not active_alive and not reserve_alive

    def swap_into_active(
        self,
        reserve_index,
        active_slot_index
    ):
        if reserve_index < 0 or reserve_index >= len(self.reserve):
            raise ValueError(
                "Invalid reserve index."
            )

        if active_slot_index < 0 or active_slot_index >= len(
            self.active_slots
        ):
            raise ValueError(
                "Invalid active slot."
            )

        reserve_card = self.reserve[reserve_index]

        if not reserve_card.is_alive:
            raise ValueError(
                "Can't field a defeated card."
            )

        if self.active_slots[active_slot_index] is not None:
            raise ValueError(
                "That active slot isn't empty."
            )

        self.reserve.pop(reserve_index)

        reserve_card.zone = "active"

        self.active_slots[active_slot_index] = reserve_card

        return reserve_card

    def handle_defeat(self, card):
        active_index = -1

        try:
            active_index = self.active_slots.index(card)
        except ValueError:
            pass

        if active_index != -1:
            self.active_slots[active_index] = None

        reserve_index = -1

        try:
            reserve_index = self.reserve.index(card)
        except ValueError:
            pass

        if reserve_index != -1:
            self.reserve.pop(reserve_index)

        card.zone = "graveyard"

        if card not in self.graveyard:
            self.graveyard.append(card)

    def has_empty_active_slot(self):
        return any(
            card is None
            for card in self.active_slots
        )